src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -8380,6 +8380,46 @@ internal static string HandleDoubleAndSingleQuote(ref string wordToComplete)
             return quote;
         }
 
+        /// <summary>
+        /// Enumerates quoted and unquoted completion text.
+        /// This makes it easier to handle different variations of completions with consideration of quotes.
+        /// </summary>
+        /// <param name="wordToComplete">The word to complete.</param>
+        /// <param name="possibleCompletionValues">The possible completion values to iterate.</param>
+        /// <param name="toolTipMapping">The optional tool tip mapping delegate.</param>
+        /// <returns></returns>
+        internal static IEnumerable<CompletionResult> EnumerateQuotedAndUnquotedCompletionText(
+            string wordToComplete,
+            IEnumerable<string> possibleCompletionValues,
+            Func<string, string> toolTipMapping = null)
+        {
+            string quote = HandleDoubleAndSingleQuote(ref wordToComplete);
+            var pattern = WildcardPattern.Get(wordToComplete + "*", WildcardOptions.IgnoreCase);
+
+            foreach (string value in possibleCompletionValues)
+            {
+                if (pattern.IsMatch(value))
+                {
+                    string completionText = quote == string.Empty
+                        ? value
+                        : quote + value + quote;
+
+                    if (toolTipMapping is not null)
+                    {
+                        yield return new CompletionResult(
+                            completionText,
+                            listItemText: value,
+                            resultType: CompletionResultType.ParameterValue,
```

Comment:
I wonder if completion result type should be parameterised?

Reply:
Good catch! We need review every code path we change and parameterize if needed.

Reply:
Fixed. I kept default as `CompletionResultType.Text`. I think this is a good default because it makes refactoring some of the older completers to use this method easier since they use it. We can override it as needed and not include a breaking change.